BoxContentSDK/BoxContentSDK/Operations/BOXAPIOAuth2ToJSONOperation.m

```diff
@@ -118,16 +118,14 @@ - (BOOL)shouldErrorTriggerLogout:(NSError *)error
         BOXOAuth2Session *session = (BOXOAuth2Session *)self.session;
         
         if ([session isKindOfClass:[BOXOAuth2Session class]]) {
-            NSString *jsonString = [[BOXOAuth2Session keychainItemWrapperForUserWithID:session.user.modelID] objectForKey:(__bridge id)kSecValueData];
-            NSError *error = nil;
-            NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding]
-                                                                       options:0
-                                                                         error:&error];
-            NSString *keychainAccessToken = dictionary[keychainAccessTokenKey];
-            
-            // If our keychain token and token in memory are out of sync, we hold off on the logout.
-            if (keychainAccessToken.length > 0 && ![keychainAccessToken isEqualToString:self.accessToken]) {
-                [session restoreSessionWithKeyChainDictionary:dictionary];
+            // We're probably here because we're trying to refresh our access token right around the time it is being
+            // refreshed in another process, and we lost the race. Check the keychain, and only report refresh failure
+            // if it wasn't updated.
+            // Pause briefly to allow time for the keychain to be updated (in case we have really unlucky timing).
+            // This only happens when racing to refresh our token, which is infrequent enough that this delay should
+            // be unnoticable to a user.
+            usleep(0.3 * USEC_PER_SEC);
```

Comment:
Generally not a fan of the sleep, but if this is where we're headed, can we make it so that we only sleep if we discover that the tokens in memory match the keychain.

Reply:
Another thing to consider is that If we have multiple calls to shouldErrorTriggerLogout all converging into one thread, these usleeps could stack up. And it could get pretty bad if these are firing on the main thread.

Reply:
Added initial keychain read. These can't stack up, and can't happen on the main thread. There is a dedicated concurrent queue for network operations, where this happens, and token refresh blocks the queue, so it only happens once per token expiration.